lib/galaxy/webapps/galaxy/api/__init__.py

```diff
@@ -19,17 +20,32 @@
     model,
 )
 from galaxy.app import UniverseApplication
-from galaxy.exceptions import AdminRequiredException
+from galaxy.exceptions import (
+    AdminRequiredException,
+    UserCannotRunAsException,
+    UserInvalidRunAsException,
+)
 from galaxy.managers.jobs import JobManager
 from galaxy.managers.session import GalaxySessionManager
 from galaxy.managers.users import UserManager
 from galaxy.model import User
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.security.idencoding import IdEncodingHelper
 from galaxy.web.framework.decorators import require_admin_message
 from galaxy.work.context import SessionRequestContext
 
 
-def get_app() -> UniverseApplication:
-    return cast(UniverseApplication, galaxy_app.app)
+def get_app() -> Generator[UniverseApplication, None, None]:
+    app = cast(UniverseApplication, galaxy_app.app)
+    try:
+        app.model.set_local_session()
```

Comment:
I think we should do this as middleware instead so avoid circular dependency issues and such and needing this to be "depended on" before anything else. That approach would also leverage more standard Python webapp patterns and depends less of specifics of FastAPI dependency injection stuff also. Maybe I'm wrong though and that won't be a problem. I just hit a relatively subtle issue when rebasing #11180 on top of this.

Reply:
The reason I didn't use a middleware is that they run on every request, so would also run on future async routes or routes that don't need the database at all. Maybe those should be entirely separated apps, in that case we could omit the sync middleware? I don't mind making this a middleware for now.

 I think the reason this doesn't feel awesome is that we don't have an isolated `get_db` that doesn't depend on app. Still a hack, but maybe better, would be moving this into `get_model` ?

Is this problem you saw https://github.com/galaxyproject/galaxy/pull/11180#issuecomment-780847940 ? In which case having a middleware would still give you the same issue, and one possible fix is https://github.com/galaxyproject/galaxy/pull/11180#discussion_r578278085